### PR TITLE
Windows instances under AWS

### PR DESCRIPTION
**Pull request by ken9ross**
## Intent of Change
- New feature (non-breaking change which adds functionality)

## Description
Add support for Windows instances in AWS

## Motivation and Context
Capability building
part of #299 

### Prerequisite PRs:

- None

### Dependent PRs:

- None

### Consumer Actions:

- New build profile of awswin similar to awslinux2

### CODE REVIEW

File: aws/extensions/computetask_awscli/extension.ftl [hunk 1]
```diff
@@ -26,6 +26,9 @@
 
     [#local content = {}]
     [#switch operatingSystem.Family ]
+        [#case "windows" ]
+        [#break]
+
```
Comment (roleyfoley) [inline aws/extensions/computetask_awscli/extension.ftl:31]: Are we not installing the aws cli? or is it already bundled?

File: aws/extensions/computetask_awswin_cfninit/extension.ftl [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    ]
+    supportedTypes=[
+        EC2_COMPONENT_TYPE,
+        ECS_COMPONENT_TYPE,
+        COMPUTECLUSTER_COMPONENT_TYPE,
+        BASTION_COMPONENT_TYPE
+    ]
+    scopes=[
+        COMPUTETASK_EXTENSION_SCOPE
+    ]
```
Comment (roleyfoley) [inline aws/extensions/computetask_awswin_cfninit/extension.ftl:19]: This now needs to include handling of a second cfn-init pass for when we need to handle cloudformation actions that depend on the ec2 instance

Comment (roleyfoley) [reply aws/extensions/computetask_awswin_cfninit/extension.ftl:19]: Actually looks like you've added another function for that. This one can be removed. We either have wait or autoscale group now

File: aws/extensions/computetask_awswin_ecs/extension.ftl [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+            [#local userInit = mergeObjects(
+                                userInit,
+                                {
+                                    userName : {
+                                        "groups" : [ "docker" ],
+                                        "uid" : details.UID
+                                    }
+                                })]
+        [/#list]
+    [/#if]
```
Comment (roleyfoley) [inline aws/extensions/computetask_awswin_ecs/extension.ftl:50]: Might need to review this and see what the windows config is for user provisioning.

File: aws/extensions/computetask_awswin_ospatching/extension.ftl [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    [#local OSPatching = _context.InstanceOSPatching ]
+    [#local schedule = OSPatching.Schedule ]
+    [#local securityOnly = OSPatching.SecurityOnly ]
+
+    [#local updateCommand = "yum clean all && yum -y update"]
+
+<!-- Windows instances are patched by new AMIs
+    [#if OSPatching.Enabled ]
+ -->
```
Comment (roleyfoley) [inline aws/extensions/computetask_awswin_ospatching/extension.ftl:32]: The awslinux instances are as well. The AMIs are released in monthly cycles I think. so could miss some updates that have been released in between. the AMI cycle

File: aws/inputseeders/aws/masterdata.json [hunk 5]
```diff
@@ -946,6 +956,127 @@
                     }
                 }
             }
+        },
+        "_awswin" : {
```
Comment (roleyfoley) [inline aws/inputseeders/aws/masterdata.json:960]: I think this is probably better as windows rather than awswin. awslinux is an actual distribution. Where the Windows install is really windows.

File: aws/inputseeders/aws/masterdata.json [hunk 6]
```diff
@@ -946,6 +956,127 @@
...
+                                "_computetask_awswin_vpc_lb",
+                                "_computetask_awswin_notavailable"
+                            ]
+                        }
+                    },
+                    "computecluster" : {
+                        "ComputeInstance" : {
+                            "OperatingSystem" : {
+                                "Family" : "windows",
+                                "Distribution" : "awswin",
```
Comment (roleyfoley) [inline aws/inputseeders/aws/masterdata.json:998]: I don't think this is a real distribution. We should reserve this for when its required.

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: aws/extensions/computetask_awscli/extension.ftl (hunk 1 above)
Comment (ken9ross) [reply aws/extensions/computetask_awscli/extension.ftl:31]: Doco says AWS CLI is already bundled

File: aws/inputseeders/aws/masterdata.json (hunk 6 above)
Comment (ken9ross) [reply aws/inputseeders/aws/masterdata.json:998]: Since it's not a vanilla windows install I was appropriating the term "distribution" to describe the AWS customised windows install (AWS CLI, EC2Lauch etc). It's not as if there is (or likely ever will be - would need MS to opensource Windows) a Redhat Windows and a MS Windows, unlike Linux.

The other potential "distribution" I thought about was the "preview" windows versions from AWS. I was thinking of awswin and awspwin, but they could be "preview" and "ga" distributions just as easily. Thoughts?

File: aws/extensions/computetask_awswin_ecs/extension.ftl (hunk 3 above)
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ecs/extension.ftl:50]: Sorry yes, ecs is currently junk and a placeholder for me

**Comment by ml019** [comment]
@ken9ross the profile refactoring has been merged so you should be good to rebase and maybe move this to review?

File: aws/extensions/computetask_awswin_hamletenv/extension.ftl [hunk 7]
```diff
@@ -66,6 +66,11 @@
         priority=2
         engine=AWS_EC2_CFN_INIT_COMPUTE_TASK_CONFIG_TYPE
         content={
+                "packages" : {
+                    "msi" : {
+                        "awscli": "https://awscli.amazonaws.com/AWSCLIV2.msi"
+                    }
```
Comment (roleyfoley) [inline aws/extensions/computetask_awswin_hamletenv/extension.ftl:72]: Should we move this to the awscli step? then its handled as part of the same extension?

Comment (ken9ross) [reply aws/extensions/computetask_awswin_hamletenv/extension.ftl:72]: Good point - done

File: aws/inputseeders/aws/masterdata.json [hunk 8]
```diff
@@ -648,6 +693,23 @@
...
+            "TimeFormat": "%Y/%m/%d %H:%M:%S%Z:"
+        },
+        "/cwa/amazon-cloudwatch-agent.log": {
+            "FilePath": "/cwa/amazon-cloudwatch-agent.log",
+            "TimeFormat": "%H: %M: %S%y%b%-d"
+        },
+        "/awsssm/amazon-ssm-agent.log": {
+            "FilePath": "/awsssm/amazon-ssm-agent.log",
+            "TimeFormat": "%Y-%m-%d %H:%M:%S",
+            "MultiLinePattern": "^INFO"
```
Comment (roleyfoley) [inline aws/inputseeders/aws/masterdata.json:712]: Are these the windows logs? Lest make them Windows formatted instead of something hard to follow.

File: aws/inputseeders/aws/masterdata.json [hunk 9]
```diff
@@ -952,6 +1048,176 @@
...
+            "Modes" : {
+                "*" : {
+                    "ecs" : {
+                        "Profiles" : {
+                            "Storage" : "ecs_awswin",
+                            "LogFile" : "awswinlog"
+                        },
+                        "AutoScaling": {
+                            "StartupTimeout": "20M"
+                        },
```
Comment (roleyfoley) [inline aws/inputseeders/aws/masterdata.json:1062]: 20 minutes for each box? seems quite long.

File: aws/services/cw/policy.ftl [hunk 10]
```diff
@@ -17,7 +17,18 @@
...
+            [
+                getPolicyStatement(
+                    [
+                        "cloudwatch:PutMetricData",
+                        "ec2:DescribeTags"
+                    ],
+                    "*")
+            ],
+            []
+        )
```
Comment (roleyfoley) [inline aws/services/cw/policy.ftl:31]: I don't think these are OS dependent.  the linux instances can put metric data as well. We just didn't have it setup. Probably better to make this standard permissions on cwLogs across the board

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 4 above)
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: Not sure this is the same for Windows as patches are released on "Patch Tuesday" (ie monthly) and the AMIs are updated with 5 working days thereafter. I'm not aware of people using automated patching on Windows servers, but I'll have a dig through Windows Update

File: aws/inputseeders/aws/masterdata.json (hunk 8 above)
Comment (ken9ross) [reply aws/inputseeders/aws/masterdata.json:712]: Yes and no. The FilePath is used to identify the windows log and also the logstream name, hence using a function to translate between them. Would it be better to create a second parameter?

File: aws/inputseeders/aws/masterdata.json (hunk 9 above)
Comment (ken9ross) [reply aws/inputseeders/aws/masterdata.json:1062]: The doco lists up to 15 minutes for a box to start and 15 minutes to join a cluster. If I've read the doco aright, that's a maximum time for a response (not a minimum), correct?

File: aws/extensions/computetask_awswin_cwlog/extension.ftl [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+        ECS_COMPONENT_TYPE,
+        COMPUTECLUSTER_COMPONENT_TYPE,
+        BASTION_COMPONENT_TYPE
+    ]
+    scopes=[
+        COMPUTETASK_EXTENSION_SCOPE
+    ]
+/]
+
+[#function windows_path_converter nix_path ]
```
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_cwlog/extension.ftl:22]: See below comment about FilePath being used to identify the windows log and also the logstream name

File: aws/extensions/computetask_awswin_ospatching/extension.ftl [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    ]
+/]
+
+[#macro shared_extension_computetask_awswin_ospatching_deployment_computetask occurrence ]
+
+    [#local OSPatching = _context.InstanceOSPatching ]
+    [#local schedule = OSPatching.Schedule ]
+    [#local securityOnly = OSPatching.SecurityOnly ]
+
+    [#local updateCommand = "<script>rem Windows no update command - yum clean all && yum -y update</script>"]
```
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl:28]: nix functionality that was not translated to Windows. Little more than a memory jogger as to what wasn't translated in a comment

File: aws/services/cw/policy.ftl (hunk 10 above)
Comment (ken9ross) [reply aws/services/cw/policy.ftl:31]: Cool, was just trying to constrain my changes to avoid breaking something out of ignorance

File: aws/extensions/computetask_awswin_volumemount/extension.ftl [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+        [/#list]
+    [/#if]
+
+    [#local diskId = -1]
+    [#list volumes as id,volume ]
+
+        [#local deviceId = ""]
+        [#local osMount = ""]
+
+        [#local diskId = diskId + 1]
```
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl:79]: Joys of learning a new language - missing the obvious

But a fix in volume mounting, makes this moot - it needs to be restricted to enabled volumes, so there was an introduced bug - now fixed

File: aws/extensions/computetask_awswin_ospatching/extension.ftl [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+                    },
+                    {
+                        "DailyUpdates" : {
+                            "command" : '<script>echo \"${schedule} ${updateCommand} >> c:\\ProgramData\\Hamlet\\Logs\\update.log\" >crontab.txt && crontab crontab.txt</script>',
+                            "ignoreErrors" : false
+                        }
+                    }
+                )
+            }]
+    [#else]
```
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl:54]: Yes. I left it in as a placeholder for possible taskscheduler integrations if required

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 4 above)
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: Lets leave it for now and instead add a warning or fatal message that if OSPatching is Enabled on the solution configuration let the user know at template generation time that the standard OS patching rules apply. 

If someone wants to enable patching they can add their own extension

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 12 above)
[the remark this answers is not in the document]
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl:28]: Lets take it out its confusing for someone picking this up if they need to work with it. At the moment it just looks like a bug.

File: aws/inputseeders/aws/masterdata.json (hunk 8 above)
Comment (roleyfoley) [reply aws/inputseeders/aws/masterdata.json:712]: Might be better to include the full path. I.e C:/ etc 
Otherwise this can't be customised by people. What would happen if I want to log something on the D: Drive in this case? 

Looking at https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-logs-logstream.html 
`:` is not allowed as part of the name so we could just do a replace when setting the logstream to remove that 


C/ProgramFiles/ is pretty understandable

File: aws/inputseeders/aws/masterdata.json (hunk 9 above)
Comment (roleyfoley) [reply aws/inputseeders/aws/masterdata.json:1062]: Its more the impact on a system if the box fails to start correctly. 20 minutes to wait for an instance to timeout when it didn't start properly is a pretty big wait during scaling activities.

File: aws/services/cw/policy.ftl (hunk 10 above)
Comment (roleyfoley) [reply aws/services/cw/policy.ftl:31]: Its just that this update isn't conditional on if you are using windows or not. Its more conditional on if you are using metrics as part of your cloudwatch agent configuration. Along with that we probably need to apply some conditions to the PutMetricData action as it's a write action. If I had multiple products in this account this permission could put data to any metric

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 12 above)
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: Cool, done

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 4 above)
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: Cool, done

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 14 above)
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: Removed

File: aws/inputseeders/aws/masterdata.json (hunk 9 above)
Comment (ken9ross) [reply aws/inputseeders/aws/masterdata.json:1062]: If an ECS has a data volume it needs to do an NTFS format plus doco warns up to 15 minutes to join the cluster and can be successful

I couldn't locate how long nix was expected to take to join a cluster, but given the above it didn't seem too much to double it to 20M. Thoughts?

The risk point is an ECS that doesn't positively succeed nor positively fail but doesn't start correctly, correct?

Comment (roleyfoley) [reply aws/inputseeders/aws/masterdata.json:1062]: Ok sounds reasonable then. Just wanted to make sure that some of the dev work wasn't bleeding into the default setup.

File: aws/extensions/computetask_awswin_volumemount/extension.ftl (hunk 13 above)
[the remark this answers is not in the document]
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl]: How will this affect enabling and disabling volumes? Would it still be better to use the index value and just have gaps in the Id format?

File: aws/services/cw/policy.ftl (hunk 10 above)
Comment (ken9ross) [reply aws/services/cw/policy.ftl]: I hadn't considered products. Logically I should be restricting it base on product and environment, but I don't see how to set that through the permissions ARN. Can you point me in the right direction?

File: aws/inputseeders/aws/masterdata.json (hunk 8 above)
Comment (ken9ross) [reply aws/inputseeders/aws/masterdata.json:712]: Cool, done

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 4 above)
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: I'm getting this warning even though I've specified OSPatching/Enabled:false in masterdata for the _awswin DeploymentProfile. Any advice on what I've done wrong?

File: aws/services/cw/policy.ftl (hunk 10 above)
Comment (roleyfoley) [reply aws/services/cw/policy.ftl]: This is the one you want 
https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/iam-cw-condition-keys-namespace.html 

So either there is a namespace created for each product/env/segment or we limit it to a single namespace where all ec2 instances log their data under different dimensions

File: aws/inputseeders/aws/masterdata.json (hunk 8 above)
Comment (roleyfoley) [reply aws/inputseeders/aws/masterdata.json]: not quite what I was thinking 

For the FilePath It should be the real OS path. We just do the removal of the : when the logstream property is set
Then the user understands and enters a real path and we handle the weirdness of AWS underneath. With the configuration as it is we are now showing the weirdness to users and making them have to understand it

File: aws/extensions/computetask_awswin_volumemount/extension.ftl (hunk 13 above)
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl]: It doesn't, unless I'm misunderstanding how that works. Windows doesn't use the volume names but the incremental number of the volume (based on the order of the enabled volumes, so I have to make sure the numbering aligns is all. eg
```
                "Volumes": {
                    "root" : {
                        "Device" : "/dev/sda1",
                        "Size" : 40
                    },
                    "userdata" : {
                        "Enabled" : false,
                        "Device" : "xvdd",
                        "Size" : 50,
                        "MountPath" : "D"
                    },
                    "data": {
                        "Enabled" : true,
                        "Device": "xvdz",
                        "Size": 10,
                        "MountPath" : "c:\\esdata"
                    }
                }
```
Device /dev/sda1 is Disk0 and device xvdz is Disk1

[the remark this answers is not in the document]
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl]: So if thats defined by the Id and I disable and enable a volume on one of my deployments what would happen?

File: aws/services/cw/policy.ftl (hunk 10 above)
Comment (ken9ross) [reply aws/services/cw/policy.ftl]: I think the piece of the puzzle that I'm missing is hidden in that last sentence. As product/env/segment are constructs of ours, how do they relate to ARNs (I'm assuming an ARN is a unique reference, correct?)

File: aws/inputseeders/aws/masterdata.json (hunk 8 above)
Comment (ken9ross) [reply aws/inputseeders/aws/masterdata.json]: Cool, done

File: aws/extensions/computetask_awswin_volumemount/extension.ftl (hunk 13 above)
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl]: Then during that execution (for example) device /dev/sda1 is Disk0, device xvdd would become Disk1 and device xvdz would be removed (assuming you set xvdd to enabled and xvdz to disabled). Disk0 and disk1 are used by diskpart during the userdata run to connect the disks when the EC2 spins up.

Does that answer your question or am I missing something?

File: aws/services/cw/policy.ftl (hunk 10 above)
Comment (roleyfoley) [reply aws/services/cw/policy.ftl]: So its within the configuration of the agent. 

https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-Configuration-File-Details.html 

within the Metrics configuration section there is the definiton of the namespace. This is the top level grouping that is used to store CloudWatch Metrics under 

So we could create a namespace that is based on the name of the component or the path of the component name. Then all instance metrics would be stored under that using the instance Id as a dimension which allows for metrics to be stored under different granularities 

Then when creating the IAM policy you would provide the Metric namespace that the role has permissions to write into

File: aws/extensions/computetask_awswin_cwlog/extension.ftl [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+                                r'                "measurement": [ ',
+                                r'                    "% Committed Bytes In Use" ',
+                                r'                ], ',
+                                r'                "metrics_collection_interval": 60 ',
+                                r'            }, ',
+                                r'            "statsd": { ',
+                                r'                "metrics_aggregation_interval": 60, ',
+                                r'                "metrics_collection_interval": 10, ',
+                                r'                "service_address": ":8125" ',
+                                r'            } ',
```
Comment (roleyfoley) [inline aws/extensions/computetask_awswin_cwlog/extension.ftl:221]: We shouldn't need statsd by default this is a service endpoint that applications can send metrics to. I'd call this an advanced feature and adding it by default might be a bir far

File: aws/extensions/computetask_awswin_cwlog/extension.ftl [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+                                r'                        "log_stream_name": "{instance_id}/security" ',
+                                r'                    } ',
+                                r'                ] ',
+                                r'            } ',
+                                r'        } ',
+                                r'    }, ',
+                                r'    "metrics": { ',
+                                r'        "append_dimensions": { ',
+                                r'            "AutoScalingGroupName": "${aws:AutoScalingGroupName}", ',
+                                r'            "ImageId": "${aws:ImageId}", ',
```
Comment (roleyfoley) [inline aws/extensions/computetask_awswin_cwlog/extension.ftl:197]: Image Id might be  a bit far here, I think just ASG, InstanceId and Type are good dimensions

**Review by roleyfoley — CHANGES_REQUESTED** [review]
A couple of extra things. 

Also why are there double \\ in all of the paths?

File: aws/extensions/computetask_awswin_cwlog/extension.ftl (hunk 15 above)
Comment (ken9ross) [reply aws/extensions/computetask_awswin_cwlog/extension.ftl:221]: Cool

File: aws/extensions/computetask_awswin_cwlog/extension.ftl (hunk 16 above)
Comment (ken9ross) [reply aws/extensions/computetask_awswin_cwlog/extension.ftl:197]: Cool

File: aws/services/cw/policy.ftl (hunk 10 above)
Comment (ken9ross) [reply aws/services/cw/policy.ftl]: Cool, thanks, done

File: aws/services/cw/policy.ftl [hunk 17]
```diff
@@ -21,7 +21,24 @@
...
+                    "cloudwatch:PutMetricData"
+                ],
+                "*",
+                "",
+                logGroupName?has_content?then(
+                    {
+                        "StringEquals": {
+                            "cloudwatch:namespace": 
+                                "CWAgent"+logGroupName?keep_before_last("/")?keep_before_last("/")
+                        }
```
Comment (roleyfoley) [inline aws/services/cw/policy.ftl:33]: This feels quite confusing and I wouldn't know the name straight off, since its used for alerts it needs to be easy to understand what the value will be 

- I think this would be better as its own macro that returns a separate policy specific for metric data pushing. And that has a namespace value. That way its formatted outside of the policy statement and can be included as part of the ec2 instance state potentially 
- We could also leave out creating a dedicated namespace for each component and rely on the dimensions provided for each ec2 instance. This would align with how metrics are collected. Then the permission applied would always been to the CWAgent namespace for these metrics and the dimensions can be filtered. That still limits the potential for overriding metrics but makes it easier to understand and aligns with how metrics are stored across the board

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: aws/extensions/computetask_awswin_volumemount/extension.ftl (hunk 13 above)
[the remark this answers is not in the document]
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl]: My question is more how flexible is the disk updates on an existing instance? Can I add and remove disks on an existing instance or will adding a new one force it to mount the disk to another location? 

I.e. if I add a disk in between root and userdata will that disk mount to the location it has been configured to use or will that use the diskId for userdata and end up on the D: drive?

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 4 above)
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: Is this still happening or was it resolved?

File: aws/extensions/computetask_awswin_cwlog/extension.ftl [hunk 18]
```diff
@@ -33,6 +33,16 @@
     [#return retval ]
 [/#function]
 
+[#function nix_path_converter win_path ]
+    [#local seqWin = win_path?split(r"\")]
+    [#if (seqWin[0])?matches('[c-zC-Z]:')]
+        [#local seqWin = [(seqWin[0])?keep_before(":")] + seqWin[1..]]
```
Comment (roleyfoley) [inline aws/extensions/computetask_awswin_cwlog/extension.ftl:39]: Why not 

path?replace(":", "") 

That would remove any reference to : which is not allowed at anypoint in the logstream name?

File: aws/extensions/computetask_awswin_cwlog/extension.ftl [hunk 19]
```diff
@@ -51,9 +61,9 @@
             [#local logContent +=
                 [
                     r'{',
-                    r'    "file_path": "' + windows_path_converter(logFileDetails.FilePath) + '",',
+                    r'    "file_path": "' + (logFileDetails.FilePath)?replace(r"\",r"\\") + '",',
```
Comment (roleyfoley) [inline aws/extensions/computetask_awswin_cwlog/extension.ftl:64]: I think the JSON output handling should cover the conversion and escaping. Was it not?

**Review by roleyfoley — CHANGES_REQUESTED** [review]

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 4 above)
Comment (ken9ross) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: Still happening

File: aws/extensions/computetask_awswin_volumemount/extension.ftl (hunk 13 above)
[the remark this answers is not in the document]
Comment (ken9ross) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl]: It won't force it to another location (since the location is explicit - ie drive letter or path - and separate from the disk number), but there will be an issue with disks as Windows doesn't use raw disks (so they have to be formatted). I'll need to check that existing disks aren't clobbered during an update

File: aws/extensions/computetask_awswin_ospatching/extension.ftl (hunk 4 above)
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_ospatching/extension.ftl]: Should now be fixed in b211b66 
It looks like the Environment Patching configuration was overriding the instance. The instance should be preferred as its the most specific

File: aws/extensions/computetask_awswin_cwlog/extension.ftl (hunk 18 above)
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_cwlog/extension.ftl]: Fixed and moved to using native object support to handle escaping and make it easier to follow the extension 

see 4a9fa5f

File: aws/extensions/computetask_awswin_cwlog/extension.ftl (hunk 19 above)
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_cwlog/extension.ftl]: Covered in 4a9fa5f

File: aws/services/cw/policy.ftl (hunk 17 above)
Comment (roleyfoley) [reply aws/services/cw/policy.ftl]: Addressed in 1bd6d75

File: aws/extensions/computetask_awswin_volumemount/extension.ftl (hunk 13 above)
[the remark this answers is not in the document]
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl]: @ken9ross do we have an answer on your checks on the disks?

[the remark this answers is not in the document]
Comment (roleyfoley) [reply aws/extensions/computetask_awswin_volumemount/extension.ftl]: Testing of this has shown that the instances are replaced when disks are added, so this means that there isn't a risk of disks being remapped and causing data to be overriden. I think we can call this resolved and work through it if it becomes an issue

**Review by roleyfoley — APPROVED** [review]
Great Work @ken9ross awesome to test out how flexible our approach is and get a good result from it
